pkg/dispatcher/dispatcher.go

```diff
@@ -0,0 +1,204 @@
+		checkInterval:            checkInterval,
+		instanceExpirationPeriod: instanceExpirationPeriod,
+		consumerSet:              mapset.NewSet[string](),
+		resyncChan:               make(chan string, 100),
+	}
+}
+
+// Start initializes and runs the DispatcherImpl instance.
+// It periodically updates the heartbeat for the current Maestro instance,
+// checks Maestro instances and consumers to maintain the consistent hash ring,
+// keeps track of consumers that belong to the current instance, and sends resync signals
+// for consumers that are added to the current instance.
+func (d *DispatcherImpl) Start(ctx context.Context) error {
+	logger := logger.NewOCMLogger(ctx)
+	instance := &api.Instance{
+		Name: d.instanceID,
+		Meta: api.Meta{
+			ID:        d.instanceID,
+			UpdatedAt: time.Now(),
+		},
+	}
+	instance, err := d.instanceDao.UpSert(ctx, instance)
+	if err != nil {
+		return fmt.Errorf("unable to create maestro instance: %s", err.Error())
+	}
+	logger.Infof("Created maestro instance %s in database", instance.Name)
+
+	instances, err := d.instanceDao.All(ctx)
+	if err != nil {
+		return fmt.Errorf("unable to list maestro instances: %s", err.Error())
+	}
+
+	// initialize the consistent hash ring
+	members := []consistent.Member{}
+	for _, instance := range instances {
+		if instance.UpdatedAt.After(time.Now().Add(time.Duration(-1*d.instanceExpirationPeriod) * time.Second)) {
+			members = append(members, instance)
+		}
+	}
+
+	d.consistent = consistent.New(members, consistent.Config{
+		PartitionCount:    7,
+		ReplicationFactor: 20,
+		Load:              1.25,
+		Hasher:            hasher{},
+	})
+	logger.Infof("Initialized consistent hash ring with members: %v", members)
+
+	consumers, err := d.consumerDao.All(ctx)
+	if err != nil {
+		return fmt.Errorf("unable to list consumers: %s", err.Error())
+	}
+
+	// initialize the consumer set for current instance
+	for _, consumer := range consumers {
+		instanceID := d.consistent.LocateKey([]byte(consumer.ID)).String()
+		if instanceID == d.instanceID {
+			// new consumer added to the current instance, need to resync resource status updates for this consumer
+			d.consumerSet.Add(consumer.ID)
+			logger.Infof("Added new consumer %s to consumer set for instance %s", consumer.ID, d.instanceID)
+			d.resyncChan <- consumer.ID
+		}
+	}
+	logger.Infof("Initialized consumers %s for current instance %s", d.consumerSet.String(), d.instanceID)
+
+	pulseTicker := time.NewTicker(time.Duration(d.pulseInterval) * time.Second)
+	checkTicker := time.NewTicker(time.Duration(d.checkInterval) * time.Second)
+	for {
+		select {
+		case <-ctx.Done():
+			pulseTicker.Stop()
+			checkTicker.Stop()
+			close(d.resyncChan)
+			return nil
+		case <-pulseTicker.C:
+			logger.Infof("Updating heartbeat for maestro instance: %s", instance.Name)
+			// update the heartbeat for current instance
+			instance.UpdatedAt = time.Now()
+			instance, err = d.instanceDao.Replace(ctx, instance)
+			if err != nil {
```

Comment:
this might need better treatment. if we only return error. Do we wait for the next pulse to update?

Reply:
Yes, does it make sense?
If the instance doesn't not pulse within the last three check interval, then it is considered as dead.